File: web/src/lib/slots.ts
// Slots and the auto-assignment / one-move rules, mirrored from the server (contract §2).
import type { PlacedElement, SizeClass, Slot } from "./api";

export const SLOTS: Record<Slot, number> = { center: 1.0, top: 0.7, bottom: 0.7, left: 0.5, right: 0.5 };
export const SLOT_ORDER: Slot[] = ["center", "top", "bottom", "left", "right"];
export const FILL_ORDER: Slot[] = ["top", "bottom", "left", "right"];

export interface Pickable {
  id: string;
  label: string;
  image_url: string | null;
  size_class?: SizeClass | null;
  origin?: PlacedElement["origin"];
  caption?: string;
}

export function mayOccupy(size: SizeClass | null | undefined, slot: Slot): boolean {
  return slot !== "center" || size === "large";
}

/**
 * Pick order preserved: the first `large` element takes center; the rest fill
 * top, bottom, left, right. No large → center stays empty (max 4).
 * Returns null for an element that does not fit (overflow).
 */
export function autoAssign(picks: Pickable[]): PlacedElement[] {
  const firstLarge = picks.find((p) => p.size_class === "large");
  const out: PlacedElement[] = [];
  if (firstLarge) out.push(toPlaced(firstLarge, "center"));
  let k = 0;
  for (const p of picks) {
    if (firstLarge && p.id === firstLarge.id) continue;
    if (k >= FILL_ORDER.length) break;
    out.push(toPlaced(p, FILL_ORDER[k++]));
  }
  return out;
}

/** How many more elements can be picked given the current picks. */
export function capacity(picks: Pickable[]): number {
  const hasLarge = picks.some((p) => p.size_class === "large");
  return (hasLarge ? 5 : 4) - picks.length;
}

export function toPlaced(p: Pickable, slot: Slot): PlacedElement {
  return {
    element_id: p.id,
    slot,
    label: p.label,
    image_url: p.image_url,
    origin: p.origin,
    size_class: p.size_class ?? null,
    caption: p.caption,
  };
}

export type MoveType = "add" | "remove" | "swap" | "move";

export interface Move {
  type: MoveType;
  element_id: string;
  to_element_id?: string;
  to_slot?: Slot;
}

export function freeSlots(elements: PlacedElement[]): Slot[] {
  const used = new Set(elements.map((e) => e.slot));
  return SLOT_ORDER.filter((s) => !used.has(s));
}

/** Apply one move for the live preview; throws a human message when illegal (mirrors the server). */
export function applyMove(elements: PlacedElement[], move: Move, lookup: (id: string) => Pickable | undefined): PlacedElement[] {
  const cur = elements.map((e) => ({ ...e }));
  const on = cur.find((e) => e.element_id === move.element_id);
  switch (move.type) {
    case "add": {
      if (on) throw new Error("that element is already on the card");
      if (cur.length >= 5) throw new Error("the card is full");
      const el = lookup(move.element_id);
      if (!el) throw new Error("unknown element");
      const free = freeSlots(cur).filter((s) => mayOccupy(el.size_class, s));
      const slot = move.to_slot ?? free[0];
      if (!slot || !free.includes(slot)) throw new Error(el.size_class === "small" ? "no free slot (small symbols cannot take the center)" : "no free slot");
      cur.push(toPlaced(el, slot));
      return cur;
    }
    case "remove": {
      if (!on) throw new Error("that element is not on the card");
      if (cur.length <= 2) throw new Error("a card keeps at least two elements");
      return cur.filter((e) => e.element_id !== move.element_id);
    }
    case "swap": {
      if (!on) throw new Error("that element is not on the card");
      if (!move.to_element_id) throw new Error("pick a replacement");
      if (cur.some((e) => e.element_id === move.to_element_id)) throw new Error("the replacement is already on the card");
      const el = lookup(move.to_element_id);
      if (!el) throw new Error("unknown element");
      if (!mayOccupy(el.size_class, on.slot)) throw new Error("a small symbol cannot take the center");
      return cur.map((e) => (e.element_id === move.element_id ? toPlaced(el, on.slot) : e));
    }
    case "move": {
      if (!on) throw new Error("that element is not on the card");
      if (!move.to_slot) throw new Error("pick a slot");
      if (!freeSlots(cur).includes(move.to_slot)) throw new Error("that slot is taken");
      if (!mayOccupy(on.size_class, move.to_slot)) throw new Error("a small symbol cannot take the center");
      return cur.map((e) => (e.element_id === move.element_id ? { ...e, slot: move.to_slot! } : e));
    }
  }
}

/** "swapped Crown for Wheel", "added Star to top" … */
export function moveWords(m: { type: MoveType; element_label?: string; element_id: string; to_element_label?: string | null; to_element_id?: string | null; to_slot?: Slot | null }): string {
  const a = m.element_label ?? m.element_id;
  const b = m.to_element_label ?? m.to_element_id ?? "";
  switch (m.type) {
    case "add":
      return `added ${a}${m.to_slot ? ` to ${m.to_slot}` : ""}`;
    case "remove":
      return `removed ${a}`;
    case "swap":
      return `swapped ${a} for ${b}`;
    case "move":
      return `moved ${a} to ${m.to_slot ?? "another slot"}`;
  }
}

File: api/pixie/slots.py
"""Five fixed slots (spec v4 §5, contract §2). A card version is 2–5 elements, one per slot.

Salience of an element on a version is its slot's value. `large` elements may go anywhere,
`small` ones never in `center`. Auto-assignment: pick order; the first `large` takes `center`,
the rest fill top, bottom, left, right; with no `large` picked, `center` stays empty (max 4).
"""
from __future__ import annotations

from typing import Mapping, Sequence

import numpy as np

SLOTS: dict[str, float] = {"center": 1.0, "top": 0.7, "bottom": 0.7, "left": 0.5, "right": 0.5}
SLOT_ORDER: list[str] = ["center", "top", "bottom", "left", "right"]
FILL_ORDER: list[str] = ["top", "bottom", "left", "right"]
MIN_ELEMENTS = 2
MAX_ELEMENTS = 5
MOVE_TYPES = ("add", "remove", "swap", "move")


def may_occupy(size_class: str | None, slot: str) -> bool:
    return slot != "center" or size_class == "large"


def salience(slot: str) -> float:
    return SLOTS[slot]


def auto_assign(elements: Sequence[Mapping]) -> list[dict]:
    """elements: dicts with `id` and `size_class`, in pick order → [{element_id, slot}]."""
    out: list[dict] = []
    center_taken = False
    fill = iter(FILL_ORDER)
    for e in elements:
        eid = e.get("id") or e.get("element_id")
        if not center_taken and e.get("size_class") == "large":
            out.append({"element_id": eid, "slot": "center"})
            center_taken = True
            continue
        slot = next(fill, None)
        if slot is None:
            raise ValueError("too many elements: without a large element in the center a card holds at most 4")
        out.append({"element_id": eid, "slot": slot})
    return out


def validate_version(elements: Sequence[Mapping], elem_by_id: Mapping[str, Mapping]) -> None:
    """Raise ValueError with a human message when a version breaks the slot rules."""
    n = len(elements)
    if n < MIN_ELEMENTS:
        raise ValueError(f"a card needs at least {MIN_ELEMENTS} elements (got {n})")
    if n > MAX_ELEMENTS:
        raise ValueError(f"a card holds at most {MAX_ELEMENTS} elements (got {n})")
    seen_e: set[str] = set()
    seen_s: set[str] = set()
    for item in elements:
        eid, slot = item.get("element_id"), item.get("slot")
        if slot not in SLOTS:
            raise ValueError(f"unknown slot {slot!r}")
        if eid not in elem_by_id:
            raise ValueError(f"unknown element {eid!r}")
        e = elem_by_id[eid]
        if not e.get("size_class"):
            raise ValueError(f"{e.get('label', eid)} is a group, not a placeable symbol")
        if eid in seen_e:
            raise ValueError(f"{e.get('label', eid)} is already on the card")
        if slot in seen_s:
            raise ValueError(f"slot {slot} is used twice")
        if not may_occupy(e.get("size_class"), slot):
            raise ValueError(f"{e.get('label', eid)} is a small symbol and cannot take the center")
        seen_e.add(eid)
        seen_s.add(slot)


def design_row(version_elements: Sequence[Mapping], leaf_pos: Mapping[str, int], E: int) -> np.ndarray:
    row = np.zeros(int(E), dtype=float)
    for item in version_elements:
        j = leaf_pos.get(item["element_id"])
        if j is not None:
            row[j] = SLOTS[item["slot"]]
    return row


def free_slots(version_elements: Sequence[Mapping]) -> list[str]:
    used = {it["slot"] for it in version_elements}
    return [s for s in SLOT_ORDER if s not in used]


def apply_move(version_elements: Sequence[Mapping], move: Mapping, elem_by_id: Mapping[str, Mapping]) -> list[dict]:
    """One move → the next version's elements. Exactly one element changes. Raises ValueError otherwise."""
    kind = move.get("type")
    if kind not in MOVE_TYPES:
        raise ValueError(f"move type must be one of {MOVE_TYPES}")
    cur = [{"element_id": it["element_id"], "slot": it["slot"]} for it in version_elements]
    by_eid = {it["element_id"]: it for it in cur}
    eid = move.get("element_id")
    if eid not in elem_by_id:
        raise ValueError(f"unknown element {eid!r}")
    e = elem_by_id[eid]
    label = e.get("label", eid)

    if kind == "add":
        if eid in by_eid:
            raise ValueError(f"{label} is already on the card")
        if len(cur) >= MAX_ELEMENTS:
            raise ValueError("the card is full; remove or swap instead")
        to_slot = move.get("to_slot")
        free = free_slots(cur)
        if to_slot is None:
            cands = [s for s in free if may_occupy(e.get("size_class"), s)]
            if not cands:
                raise ValueError(f"no free slot can take {label}")
            to_slot = cands[0]
        elif to_slot not in SLOTS:
            raise ValueError(f"unknown slot {to_slot!r}")
        elif to_slot not in free:
            raise ValueError(f"slot {to_slot} is taken")
        cur.append({"element_id": eid, "slot": to_slot})

    elif kind == "remove":
        if eid not in by_eid:
            raise ValueError(f"{label} is not on the card")
        if len(cur) <= MIN_ELEMENTS:
            raise ValueError(f"a card keeps at least {MIN_ELEMENTS} elements; swap instead")
        cur = [it for it in cur if it["element_id"] != eid]

    elif kind == "swap":
        to = move.get("to_element_id")
        if eid not in by_eid:
            raise ValueError(f"{label} is not on the card")
        if to not in elem_by_id:
            raise ValueError(f"unknown element {to!r}")
        if to in by_eid:
            raise ValueError(f"{elem_by_id[to].get('label', to)} is already on the card")
        if to == eid:
            raise ValueError("swap needs a different element")
        slot = by_eid[eid]["slot"]
        for it in cur:
            if it["element_id"] == eid:
                it["element_id"] = to

    elif kind == "move":
        to_slot = move.get("to_slot")
        if eid not in by_eid:
            raise ValueError(f"{label} is not on the card")
        if to_slot not in SLOTS:
            raise ValueError(f"unknown slot {to_slot!r}")
        if to_slot == by_eid[eid]["slot"]:
            raise ValueError(f"{label} is already in {to_slot}")
        if to_slot not in free_slots(cur):
            raise ValueError(f"slot {to_slot} is taken")
        for it in cur:
            if it["element_id"] == eid:
                it["slot"] = to_slot

    validate_version(cur, elem_by_id)
    return cur


def describe_move(move: Mapping, elem_by_id: Mapping[str, Mapping]) -> str:
    lab = lambda x: elem_by_id.get(x, {}).get("label", x)  # noqa: E731
    k = move.get("type")
    if k == "add":
        return f"add {lab(move.get('element_id'))}" + (f" ({move['to_slot']})" if move.get("to_slot") else "")
    if k == "remove":
        return f"remove {lab(move.get('element_id'))}"
    if k == "swap":
        return f"swap {lab(move.get('element_id'))} → {lab(move.get('to_element_id'))}"
    if k == "move":
        return f"move {lab(move.get('element_id'))} to {move.get('to_slot')}"
    return str(dict(move))
